### ISSUE TITLE
No event is fired before current contents are cached

### ISSUE THREAD

**Issue by tremby** [defunkt/jquery-pjax#451]
It would be very useful to be able to run some destruction logic on the page before it is cached. But at present `pjax:start` fires after `cachePush` so as far as I can tell there's no way to do this for a back/forward event.

**Comment by olemoign** [comment]
@mislav Actually, using `pjax:beforeSend` will handle `pjax()` calls but not handling popstate events means it can't really be used.

In my project (where I use dataTables, with the same needs that @tremby), I've added a `pjax:beforeCaching` event in the `onPjaxPopstate` function which works well.

**Comment by korotovsky** [comment]
Any progress is here? I use https://en.bem.info/libs/bem-core/v2.8.0/ (i-bem) library and it adds `*_js_inited` to each BEM-block when it initializes. I have a listener which listen for `pjax:end` and initializes the current container with BEM-blocks. BUT I also have to do `BEMDOM.detach()` to destruct all event listeners in old container *before* it will be cached and *after* XHR is completed. Because UI should respond while request is in progress.

Currently: I would like to get any point of extension for modifying old (current) container before caching. 
Ideally: It would be nice to get any point of extension and cache old container after XHR has been succeed.

// cc @mislav

**Comment by korotovsky** [comment]
@mislav BTW, there is no way to use `pjax:popstate` to destruct the old container because `cachePop` is also stores the current content to cache stack. And now I can't imagine how to solve it even with workaround

**Comment by tremby** [comment]
@korotovsky, you could try using my fork, mentioned above. I'm using it in several production websites, some of which are fairly high-traffic.

I just noticed, incidentally, that the relevant commit has been merged in to the heroku branch of this official repository, so it would appear it has support from the pjax team.

**Comment by tremby** [comment]
It seems I was mistaken. Not sure why it shows that way on the graph when I look at my repository locally.

![screen shot 2015-11-16 at 10 03 11](https://cloud.githubusercontent.com/assets/199635/11190208/4ceeafda-8c49-11e5-9a48-4bd1dcf19e95.png)

(Near the bottom. Weird.)
